ksql-streams/src/main/java/io/confluent/ksql/execution/streams/materialization/ks/KsLocator.java

```diff
@@ -51,27 +62,58 @@
   }
 
   @Override
-  @SuppressWarnings("deprecation")
-  public Optional<KsqlNode> locate(final Struct key) {
-    final StreamsMetadata metadata = kafkaStreams
-        .metadataForKey(stateStoreName, key, keySerializer);
-
-    if (metadata == StreamsMetadata.NOT_AVAILABLE) {
-      return Optional.empty();
+  public List<KsqlNode> locate(
+      final Struct key,
+      final RoutingFilter routingFilters
+  ) {
+    final KeyQueryMetadata metadata = kafkaStreams
+        .queryMetadataForKey(stateStoreName, key, keySerializer);
+
+    // Fail fast if Streams not ready. Let client handle it
+    if (metadata == KeyQueryMetadata.NOT_AVAILABLE) {
+      LOG.debug("KeyQueryMetadata not available for state store {} and key {}",
+                stateStoreName, key);
+      throw new MaterializationException(String.format(
+          "KeyQueryMetadata not available for state store %s and key %s", stateStoreName, key));
     }
 
-    final HostInfo hostInfo = metadata.hostInfo();
-    return Optional.of(asNode(hostInfo));
+    final HostInfo activeHost = metadata.getActiveHost();
+    final Set<HostInfo> standByHosts = metadata.getStandbyHosts();
+    LOG.debug("Before filtering: Active host {} , standby hosts {}", activeHost, standByHosts);
```

Comment:
do we have a queryId to log here as well? it might help trace a query through the routing.. if we can include it for all log statements

Reply:
No, we don't have access to the query, only `KafkaStreams`